cmake/Modules/Platform/Emscripten.cmake

```diff
@@ -270,7 +270,9 @@ if (EMSCRIPTEN_GENERATE_BITCODE_STATIC_LIBRARIES)
   message(FATAL_ERROR "EMSCRIPTEN_GENERATE_BITCODE_STATIC_LIBRARIES is not compatible with the llvm backend")
 endif()
 
-set(CMAKE_EXECUTABLE_SUFFIX ".js")
+if(NOT DEFINED CMAKE_EXECUTABLE_SUFFIX)
+  set(CMAKE_EXECUTABLE_SUFFIX ".js")
+endif()
```

Comment:
It looks like we do have tests that set CMAKE_EXECUTABLE_SUFFIX  (e.g. `test/cmake/target_html/CMakeLists.txt`) but maybe they work because they don't set it on the command line.

If that is correct then it would be good to add a command line test.

Reply:
Below we use the pattern: `if ("${CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES}" STREQUAL "")` .. is there some reason to prefer one over the other?

Reply:
> but maybe they work because they don't set it on the command line.

@sbc100 I'm pretty sure those were working because they were set as non-cache variables _after_ the call to `project()` (IIRC, toolchain file is read upon several events, one of which is calling `project()`). I.e. call to `project()`, toolchain is read, toolchain sets `CMAKE_EXEUCTABLE_SUFFIX` as non-cache variable, project's CMakeLists.txt sets `CMAKE_EXEUCTABLE_SUFFIX`, overwriting the value set by the toolchain (I have not verified this, but it's my semi-educated guess). What I have written here with `if(NOT DEFINED CMAKE_EXECUTABLE_SUFFIX)` will do the right thing for the case where the user sets `CMAKE_EXECUTABLE_SUFFIX` as a cache variable (Ex. via `-D` in the configure command).

Reply:
> is there some reason to prefer one over the other?

@sbc100 There's a behavioural difference. using if(DEFINED) checks if the variable is _defined_. A variable can be defined as an empty string.

```cmake
set(empty "")
if(DEFINED empty)
    message("var empty is defined")
endif()
unset(empty)
if(DEFINED empty)
    message("I will not print")
endif()
```

Reply:
Presumably we should be consistent withing this file though.  Checking for defined in one case and empty string in another seems inconsistent.

Reply:
where there is a semantic difference, insisting upon "consistency" for appearances seems like the wrong mindset to me. I don't know what the blocks with `if ("${CMAKE_<LANG>_IMPLICIT_INCLUDE_DIRECTORIES}" STREQUAL "")` are for, so I can't speak on what should be done there. If the point is to just not interfere if anyone else has defined a value (including defining it to an empty string), then you should use `if(DEFINED)`.

Reply:
OK, we can followup with a conversion of the existing usages of `STREQUAL ""`.  I suspect they should be using `if(DEFINED)` too.